apps/emqx_prometheus/src/emqx_prometheus.erl

```diff
@@ -73,96 +63,74 @@
 
 -define(TIMER_MSG, '#interval').
 
--record(state, {push_gateway, timer, interval}).
-
-%%--------------------------------------------------------------------
-%% update new config
-update(Config) ->
-    case
-        emqx_conf:update(
-            [prometheus],
-            Config,
-            #{rawconf_with_defaults => true, override_to => cluster}
-        )
-    of
-        {ok, #{raw_config := NewConfigRows}} ->
-            case maps:get(<<"enable">>, Config, true) of
-                true ->
-                    ok = restart();
-                false ->
-                    ok = stop()
-            end,
-            {ok, NewConfigRows};
-        {error, Reason} ->
-            {error, Reason}
-    end.
-
-start() ->
-    {_, []} = emqx_prometheus_proto_v1:start(mria_mnesia:running_nodes()),
-    ok.
-
-stop() ->
-    {_, []} = emqx_prometheus_proto_v1:stop(mria_mnesia:running_nodes()),
-    ok.
-
-restart() ->
-    ok = stop(),
-    ok = start().
-
-do_start() ->
-    emqx_prometheus_sup:start_child(?APP, emqx_conf:get([prometheus])).
-
-do_stop() ->
-    case emqx_prometheus_sup:stop_child(?APP) of
-        ok ->
-            ok;
-        {error, not_found} ->
-            ok
-    end.
+-define(HTTP_OPTIONS, [{autoredirect, true}, {timeout, 60000}]).
 
 %%--------------------------------------------------------------------
 %% APIs
 %%--------------------------------------------------------------------
 
-start_link(Opts) ->
-    gen_server:start_link({local, ?MODULE}, ?MODULE, [Opts], []).
+start_link([]) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+info() ->
+    gen_server:call(?MODULE, info).
 
 %%--------------------------------------------------------------------
 %% gen_server callbacks
 %%--------------------------------------------------------------------
 
-init([Opts]) ->
-    Interval = maps:get(interval, Opts),
-    PushGateway = maps:get(push_gateway_server, Opts),
-    {ok, ensure_timer(#state{push_gateway = PushGateway, interval = Interval})}.
+init([]) ->
+    #{interval := Interval} = opts(),
+    {ok, #{timer => ensure_timer(Interval), ok => 0, failed => 0}}.
 
+handle_call(info, _From, State = #{timer := Timer}) ->
+    {reply, State#{opts => opts(), next_push_ms => erlang:read_timer(Timer)}, State};
```

Comment:
why do we need `next_push_ms` ?

Reply:
`Info/0` function is only for debugging.  I think `next_push_ms` is more straightforward than timer ref.